ygnmi/ygnmi.go

```diff
@@ -313,25 +315,27 @@ func (w *Watcher[T]) Await() (*Value[T], error) {
 // Calling Await on the returned Watcher waits for the subscription to complete.
 // It returns the last observed value and a boolean that indicates whether that value satisfies the predicate.
 func Watch[T any](ctx context.Context, c *Client, q SingletonQuery[T], pred func(*Value[T]) error, opts ...Option) *Watcher[T] {
+	wCtx, cancel := context.WithCancel(ctx)
 	w := &Watcher[T]{
-		errCh: make(chan error, 1),
+		errCh:  make(chan error, 1),
+		cancel: cancel,
 	}
 
 	resolvedOpts := resolveOpts(opts)
-	sub, err := subscribe[T](ctx, c, q, gpb.SubscriptionList_STREAM, resolvedOpts)
+	sub, err := subscribe[T](wCtx, c, q, gpb.SubscriptionList_STREAM, resolvedOpts)
 	if err != nil {
 		w.errCh <- err
 		return w
 	}
 
-	dataCh, errCh := receiveStream[T](ctx, sub, q)
+	dataCh, errCh := receiveStream[T](wCtx, sub, q)
```

Comment:
Can the cancel just be scoped to the goroutine, with no need to store anything in the watcher. Does this work?
ctx, cancel := context.WithCancel(ctx)
dataCh, errCh := receiveStream[T](wCtx, sub, q)
go func() {
  defer cancel()
  ...
}()

That would also mean the TCP session ends even if, for whatever reason, Await() is never called.

Reply:
sure, just have to cancel in a few more place to make we don't leak the context